framework/finalizer.go

```diff
@@ -0,0 +1,149 @@
+package framework
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+
+	"github.com/cenkalti/backoff"
+	"github.com/giantswarm/microerror"
+	"k8s.io/apimachinery/pkg/api/meta"
+	"k8s.io/apimachinery/pkg/types"
+)
+
+const (
+	finalizerPrefix = "operatorkit.giantswarm.io"
+)
+
+type patchSpec struct {
+	Op    string      `json:"op"`
+	Path  string      `json:"path"`
+	Value interface{} `json:"value"`
+}
+
+func (f *Framework) addFinalizer(obj interface{}) (stopReconciliation bool, err error) {
+	patch, path, result, err := createAddFinalizerPatch(obj, f.name)
+	if err != nil {
+		return false, microerror.Mask(err)
+	}
+	if patch == nil {
+		return result, err
+	}
+	p, err := json.Marshal(patch)
+	if err != nil {
+		return false, microerror.Mask(err)
+	}
+	operation := func() error {
+		res := f.restClient.Patch(types.JSONPatchType).AbsPath(path).Body(p).Do()
+		if res.Error() != nil {
+			return microerror.Mask(res.Error())
+		}
+		return nil
+	}
+	err = backoff.Retry(operation, f.backOffFactory())
+	if err != nil {
+		return false, microerror.Mask(err)
+	}
+
+	return true, nil
+}
+
+func createAddFinalizerPatch(obj interface{}, operatorName string) (patch []patchSpec, path string, stopReconciliation bool, err error) {
+	accessor, err := meta.Accessor(obj)
+	if err != nil {
+		return nil, "", false, microerror.Mask(err)
+	}
+	if accessor.GetDeletionTimestamp() != nil {
+		return nil, "", true, nil // object has been marked for deletion, we should ignore it.
+	}
+	finalizerName := getFinalizerName(operatorName)
+	if containsFinalizer(accessor.GetFinalizers(), finalizerName) {
+		return nil, "", false, nil // object already has the finalizer.
+	}
+	patch = []patchSpec{}
+	if len(accessor.GetFinalizers()) == 0 {
```

Comment:
Special case: If we want to set the first finalizer ever on the object, we have to make sure that the `finalizers` field exists. Otherwise the patch fails.

Reply:
Would be good to add `Op: "test"` to check if `/metadata/finalizers` is still empty during adding the finalizer to avoid races.